server/storage/backend/backend.go

```diff
@@ -455,7 +456,61 @@ func (b *backend) Commits() int64 {
 }
 
 func (b *backend) Defrag() error {
-	return b.defrag()
+	err := b.defrag()
+	if err != nil {
```

Comment:
Instead of doing the restore here we could probably place it inside a `defer` in `defrag()` function? However, if we do so we need to be a bit careful about releasing the locks -> order of defered function execution is important in that case.

Reply:
Not sure why do generic error handling here.
Why not identify where `b.defrag()` generates error and then improve error handler there?

Reply:
Sure that is an option, I just thought you're looking for a generic error handling.

I already know where the error happens so we can improve that. I'll add another commit.

Reply:
Added 1fb20649b2eace00c9f17f712d5019b5917805de which addresses your comment and only handles the specific error. The e2e test is adapted as well. Manual testing and the e2e test seem to confirm that this is working.

Reply:
Looks good, can you confirm that the test reproduces the bug on the old version?

Reply:
Yes it does, the test triggers a panic in etcd I just verified that.